package object_spawner_msgs.msgs;

import "pose.proto";
import "vector3d.proto";

message SpawnRequest
{
    /* @brief Type of request */
    enum Type{
        /* @brief Spawn model in simulation */
        SPAWN   = 1;
        /* @brief Move a model in simulation */
        MOVE    = 2;
        /* @brief Remove all entities from the world */
        CLEAR   = 3;
    }

    enum ModelType{
        /* @brief Spawn sphere in simulation */
        SPHERE      = 1;
        /* @brief Spawn cylinder in simulation */
        CYLINDER    = 2;
        /* @brief Spawn cube in simulation */
        BOX        = 3;
        /* @brief Spawn custom object from sdf string */
        CUSTOM     = 4;
        /* @brief Spawn a model included in gazebo model path */
        MODEL      = 5;
        /* @brief Spawn ground plane */
        GROUND     = 6;
    }

    /* @brief Type of request */
    required Type                   type = 1;
    /* @brief Model type for spawn */
    required ModelType              model_type = 2;
    /* @brief Model name */
    optional string                 name = 3;
    /* @brief Model pose */
    optional gazebo.msgs.Pose       pose = 4;
    /* @brief Model mass */
    optional double                 mass = 5;
    /* @brief Material script uri */
    optional string                 texture_uri = 6;
    /* @brief Material script name */
    optional string                 texture_name = 7;
    /* @brief radius of a cylinder or a sphere */
    optional double                 radius = 8;
    /* @brief Length of a cylinder */
    optional double                 length = 9;
    /* @brief Dimensions of a box */
    optional gazebo.msgs.Vector3d   box_size = 10;
    /* @brief Custom object sdf string */
    optional string                 sdf = 11;
}